lib/core-integration/src/Test/Integration/Scenario/API/Shelley/Transactions.hs

```diff
@@ -617,13 +620,101 @@ spec = describe "SHELLEY_TRANSACTIONS" $ do
 
         -- Get insertion slot and out of response.
         let (_, Right apiTx) = r
-        let Just sl = view (#absoluteSlotNumber . #getApiT) <$> apiTx ^. #pendingSince
+        let Just sl = absSlotB <$> apiTx ^. #pendingSince
+
+        -- The expected expiry slot (adds the hardcoded default ttl)
+        ttl <- getTTLSlots ctx defaultTxTTL
+        let txExpectedExp = sl + ttl
+
+        -- The actual expiry slot
+        let Just txActualExp = absSlotS <$> apiTx ^. #expiresAt
+
+        -- Expected and actual are fairly close
+        slotDiff txExpectedExp txActualExp `shouldSatisfy` (< 50)
+
+    it "TRANS_TTL_02 - Custom transaction expiry" $ \ctx -> do
+        (wa, wb) <- (,) <$> fixtureWallet ctx <*> fixtureWallet ctx
+        let amt = minUTxOValue :: Natural
+        let testTTL = 42 :: NominalDiffTime
+
+        basePayload <- mkTxPayload ctx wb amt fixturePassphrase
+        let payload = addTxTTL (realToFrac testTTL) basePayload
+
+        r <- request @(ApiTransaction n) ctx
+            (Link.createTransaction @'Shelley wa) Default payload
+
+        verify r
+            [ expectSuccess
+            , expectField (#status . #getApiT) (`shouldBe` Pending)
+            , expectField #expiresAt (`shouldSatisfy` isJust)
+            ]
+
+        -- Get insertion slot and out of response.
+        let (_, Right apiTx) = r
+        let Just sl = absSlotB <$> apiTx ^. #pendingSince
 
         -- The expected expiry slot (adds the hardcoded default ttl)
-        let ttl = sl + defaultTxTTL
+        ttl <- getTTLSlots ctx testTTL
+        let txExpectedExp = sl + ttl
 
-        (view #absoluteSlotNumber <$> (apiTx ^. #expiresAt))
-            `shouldBe` Just (ApiT ttl)
+        -- The actual expiry slot
+        let Just txActualExp = absSlotS <$> apiTx ^. #expiresAt
+
+        -- Expected and actual are fairly close. Any difference should only be
+        -- due to slot rounding.
+        slotDiff txExpectedExp txActualExp `shouldSatisfy` (< 50)
+
+    it "TRANS_TTL_03 - Expired transactions" $ \ctx -> do
+        (wa, wb) <- (,) <$> fixtureWallet ctx <*> fixtureWallet ctx
+        let amt = minUTxOValue :: Natural
+
+        -- this transaction is going to expire really soon.
+        basePayload <- mkTxPayload ctx wb amt fixturePassphrase
+        let payload = addTxTTL 0.1 basePayload
```

Comment:
Sounds like this is a bit flaky :s, we should choose a time slightly larger in front of the time slot.

Reply:
Hmm yes, not good.

The TTL must be small enough that the tx expires before it can get into a block, but large enough that the node allows it into its mempool.

I need another approach to testing this which isn't flaky.

Reply:
@rvl a possible approach could be to double-spend? If we make two transactions that are using a common UTxO, both should be "accepted" in the mempool and only one would go through ultimately. Then, we can set a large-enough but still short TTL and simply wait for either one to expire. 

So, the test could be:

a) Setup a wallet with a single UTxO
b) Send funds to A with TTL t
c) Forget transaction
d) Send funds to B with TTL t

Yet, this maybe does not test what you intend to test? If we want to check that the transaction is not inserted because it is expired (and not because it is double-spending), then it doesn't work :/